flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java

```diff
@@ -447,6 +453,165 @@ public void postStop() throws Exception {
+				jobMasterConfiguration.getSlotRequestTimeout(),
+				log);
+		} catch (JobExecutionException | JobException e) {
+			return FutureUtils.completedExceptionally(
+				new JobModificationException("Could not create rescaled ExecutionGraph.", e));
+		}
+
+		// 3. disable checkpoint coordinator to suppress subsequent checkpoints
+		final CheckpointCoordinator checkpointCoordinator = executionGraph.getCheckpointCoordinator();
+		checkpointCoordinator.stopCheckpointScheduler();
+
+		// 4. take a savepoint
+		final CompletableFuture<String> savepointFuture = triggerSavepoint(
+			jobMasterConfiguration.getTmpDirectory(),
+			timeout);
+
+		final CompletableFuture<ExecutionGraph> executionGraphFuture = savepointFuture
+			.thenApplyAsync(
+				(String savepointPath) -> {
+					try {
+						newExecutionGraph.getCheckpointCoordinator().restoreSavepoint(
+							savepointPath,
+							false,
+							newExecutionGraph.getAllVertices(),
+							userCodeLoader);
+					} catch (Exception e) {
+						disposeSavepoint(savepointPath);
+
+						throw new CompletionException(new JobModificationException("Could not restore from temporary rescaling savepoint.", e));
+					}
+
+					// delete the savepoint file once we reach a terminal state
+					newExecutionGraph.getTerminationFuture()
+						.whenCompleteAsync(
+							(JobStatus jobStatus, Throwable throwable) -> disposeSavepoint(savepointPath),
+							scheduledExecutorService);
+
+					return newExecutionGraph;
+				}, scheduledExecutorService)
+			.exceptionally(
+				(Throwable failure) -> {
+					// in case that we couldn't take a savepoint or restore from it, let's restart the checkpoint
+					// coordinator and abort the rescaling operation
+					if (checkpointCoordinator.isPeriodicCheckpointingConfigured()) {
+						checkpointCoordinator.startCheckpointScheduler();
+					}
+
+					throw new CompletionException(failure);
+				});
+
+		// 5. suspend the current job
+		final CompletableFuture<JobStatus> terminationFuture = executionGraphFuture.thenComposeAsync(
+			(ExecutionGraph ignored) -> {
+				currentExecutionGraph.suspend(new FlinkException("Job is being rescaled."));
+				return currentExecutionGraph.getTerminationFuture();
+			},
+			getMainThreadExecutor());
+
+		final CompletableFuture<Void> suspendedFuture = terminationFuture.thenAccept(
+			(JobStatus jobStatus) -> {
+				if (jobStatus != JobStatus.SUSPENDED) {
+					final String msg = String.format("Job %s rescaling failed because we could not suspend the execution graph.", jobGraph.getName());
+					log.info(msg);
+					throw new CompletionException(new JobModificationException(msg));
+				}
+			});
+
+		// 6. resume the new execution graph from the taken savepoint
+		final CompletableFuture<Acknowledge> rescalingFuture = suspendedFuture.thenCombineAsync(
+			executionGraphFuture,
+			(Void ignored, ExecutionGraph restoredExecutionGraph) -> {
+				// check if the ExecutionGraph is still the same
+				//noinspection ObjectEquality
```

Comment:
I think on this line there is no warning to suppress.

Reply:
There is. The `a == b` causes this. But here we want to check for referential identity.

Reply:
ok